First draft of 21.0.0.3 beta (#1087)
OpenLiberty/blogs: posts/2021-02-19-mpcontextpropagation-requesttiming-21003-beta.adoc

---
layout: post

> title: "RequestTiming 1.0, MicroProfile Context Propagation 1.2, enhancments to App Manager configuration and Liberty connection management in Open Liberty 21.0.0.3-beta"

Reviewer: ```suggestion
title: "MicroProfile Context Propagation 1.2, RequestTiming 1.0 and JakartaMail 2.0 in Open Liberty 21.0.0.3-beta"
```

Author: Done

blog_description: Open Liberty 21.0.0.3-beta contains four All Beta features (new `expandLocation` property for App Manager configuration, Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature (RequestTiming 1.0), MicroProfile Context Propagation 1.2 and Automatic Cleanup of Leaked Connections) and one feature in Jakarta EE 9 (JakartaMail). Also in 21.0.0.3-beta the link:https://github.com/OpenLiberty/open-liberty/issues/15649[issue in MicroProfile Fault Tolerance 3.0] from 21.0.0.2-beta has now been resolved meaning that the `mpFaultTolerance-3.0` feature and the `microProfile-4.0` convenience feature can be used in the Open Liberty 21.0.0.3-beta.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= RequestTiming 1.0, MicroProfile Context Propagation 1.2, enhancments to App Manager configuration and Liberty connection management in Open Liberty 21.0.0.3-beta
Jakub Pomykala <https://github.com/jakub-pomykala>
:imagesdir: /
:url-prefix:
:url-about: /

> Open Liberty 21.0.0.3-beta contains four All Beta features (new `expandLocation` property for App Manager configuration, Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature (RequestTiming 1.0), MicroProfile Context Propagation 1.2 and Automatic Cleanup of Leaked Connections) and one feature in Jakarta EE 9 (JakartaMail). Also in 21.0.0.3-beta, the link:https://github.com/OpenLiberty/open-liberty/issues/15649[issue in MicroProfile Fault Tolerance 3.0] from 21.0.0.2-beta has now been resolved meaning that the `mpFaultTolerance-3.0` feature and the `microProfile-4.0` convenience feature can be used in the Open Liberty 21.0.0.3-beta.

Reviewer: ```suggestion
Open Liberty 21.0.0.3-beta contains four new "All Beta" features and one new Jakarta EE 9 feature. We have also resolved the link:https://github.com/OpenLiberty/open-liberty/issues/15649[MicroProfile Fault Tolerance 3.0 issue] from 21.0.0.2-beta which means that the `mpFaultTolerance-3.0` feature and the `microProfile-4.0` convenience feature can be properly used in the Open Liberty 21.0.0.3-beta.
```

Author: Changed line 19 to "Open Liberty 21.0.0.3-beta contains four new "All Beta" features and one new Jakarta EE 9 feature. Also in 21.0.0.3-beta, all the MicroProfile 4.0 features are now available."

Author: @tevans78

Reviewer: ```suggestion
In addition to all Jakarta EE 9 and MicroProfile 4.0 beta features, Open Liberty 21.0.0.3-beta also contains five brand new features.
```

Author: Done

We have two beta packages for link:{url-about}[Open Liberty]:

* <<allbeta, All Beta Features>>: a larger package that contains all Open Liberty beta features (including Jakarta EE 9 beta features) and GA features and functions.
* <<jakarta, Jakarta EE 9 Beta Features>>: a lightweight package that contains only the Jakarta EE 9 features.

This means that you can now try out our in-development Open Liberty features by just adding the relevant coordinates to your build tools.

> If you try either package, <<feedback, let us know what you think>>.

Reviewer: Move this section up to the top

Author: Done

[#allbeta]
== All Beta Features package

The All Beta Features package includes the following beta features:

> * <<expandLocation, expandLocation property for App Manager configuration>>
> * <<requestTiming, Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature>>
> * <<MPContext, MicroProfile Context Propagation 1.2>>
> * <<leakedConnections, Automatic Cleanup of Leaked Connections>>

Reviewer: Re-ordered these so they should be re-ordered below to match as well
```suggestion
* <<MPContext, MicroProfile Context Propagation 1.2>>
* <<requestTiming, Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature>>
* <<leakedConnections, Automatic Cleanup of Leaked Connections>>
* <<expandLocation, expandLocation property for App Manager configuration>>
```

Author: Done

These features also join MicroProfile 4.0 features which can be found link:https://openliberty.io/blog/2021/01/26/ee9-messaging-security-21002-beta.html#allbeta[here].

[#expandLocation]
=== expandLocation property for App Manager configuration

This enhancement allows the user to specify a expansion location (`expandLocation`) on the `applicationManager` configuration to be utilized when the `autoExpand` attribute is set to "true". As currently implemented, when an application is autoExpanded the default location for the expanded files are hard coded to `${server.config.dir}/apps/expanded/`.

The user is now able to configure that location to a new value on the file system. 

> Thus for example:

Reviewer: Remove the word "Thus" ... "For example" is fine

Author: Done

[source, xml]
----
<applicationManager autoExpand="true" expandLocation="${server.config.dir}/myApps/" />
----

would result in the application being expanded at `${server.config.dir}/myApps/{appname}/`

This enhancement gives users more flexibility regarding the location of their expanded applications.

> You can find out more at link:{url-prefix}/docs/20.0.0.12/reference/config/applicationManager.html[Open Liberty applicationManager documentation]

Reviewer: Change the link to be 21.0.0.1?

Reviewer: Is there a "latest" link?

Author: Done

Author: Changed to `latest`

[#requestTiming]
=== Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature

> The Request Timing feature (`requestTiming-1.0`) provides diagnostic information when the duration of any request exceeds the configured threshold. It provides a way to monitor requests with respect to time, where it can automatically detect slow and hung requests all while providing detailed information back, such as, warning messages, thread stacks, and the creation of thread dumps for hung requests, to help diagnose the cause of the slow or hung request, in Open Liberty.

Reviewer: ```suggestion
The Request Timing feature (`requestTiming-1.0`) provides diagnostic information when the duration of any request exceeds the configured threshold. It provides a way to monitor requests with respect to time. The feature can automatically detect slow and hung requests and provide detailed diagnostic information; warning messages, thread stacks, and the creation of thread dumps.
```

Author: Done

> When a hung request is detected in the Request Timing feature, a warning message is written in the messages log file along with a dump of the tree of events that happened during the request. Following that, a set of three thread dumps will be initiated, 1 minute apart. After the completion of the three thread dumps, further set of three thread dumps are created only if new requests are detected to be hanging.

Reviewer: ```suggestion
When a hung request is detected in the Request Timing feature, a warning message is written in the messages log file along with a dump of the events that happened during the request. Following that, a set of three thread dumps will be initiated, 1 minute apart. After the completion of the three thread dumps, further set of three thread dumps are created only if new requests are detected to be hanging.
```

Author: Done

> Some operations teams do not want thread dumps to be generated, whenever a hung request is detected in the requestTiming-1.0 feature, because there will be unwanted java thread dumps for requests that are known to be normally long and the process of generating them will impact performance. In previous Open Liberty releases, there was no option to disable the thread dumps from being generated each time a hung request is detected.

Reviewer: ```suggestion
Some operations teams do not want so many thread dumps to be generated due to the performance overhead for requests that are known to be normally long. In previous Open Liberty releases, there was no option to disable the thread dumps from being generated.
```

Author: Done

> You can now control whether the Request Timing feature collects thread dumps dumps or not when a hung request is detected. By simply configuring the **NEW** `enableThreadDumps` Request Timing server configuration attribute, to false, thread dumps will not be created during hung requests. Thread dumps will be created, by default, if the new server configuration attribute is set to true or not specified at all.

Reviewer: ```suggestion
You can now control whether the Request Timing feature collects thread dumps. By setting the **NEW** `enableThreadDumps` Request Timing server configuration attribute to false, thread dumps will not be created during hung requests. If the new server configuration attribute is set to true or not specified at all, thread dumps will still be created.
```

Author: Done

   
The new Request Timing server configuration attribute can be configured in your server.xml as follows:

[source, xml]
----
<requestTiming includeContextInfo="true" slowRequestThreshold="120s" hungRequestThreshold="10s" sampleRate="1" enableThreadDumps="false"></requestTiming>`
----

The `enableThreadDumps` server configuration attribute can also be used in embedded Request Timing sub-elements: 
`<servletTiming/>` or `<jdbcTiming/>`, as follows:

[source, xml]
----
<requestTiming includeContextInfo="true" slowRequestThreshold="120s" hungRequestThreshold="10s" sampleRate="1">
    <servletTiming appName="MyApp" servletName="MyServletApp" slowRequestThreshold="100s" hungRequestThreshold="5s" enableThreadDumps="false"/>
</requestTiming>`
----

NOTE: An embedded `<servletTiming/>` or `<jdbcTiming/>` configuration in the server.xml file overrides the configured slow and hung request threshold that are defined in `<requestTiming/>`. 

For more information on the Request Timing feature, please refer to the following documentations:
- link:{url-prefix}/docs/21.0.0.1/reference/feature/requestTiming-1.0.html[Open Liberty Documentation on requestTiming-1.0]

> - {url-prefix}/docs/21.0.0.1/reference/config/requestTiming.html

Reviewer: remove me

Author: Done, not removed but made into an actual link

[MPContext]
=== MicroProfile Context Propagation 1.2

> MicroProfile Context Propagation is a standalone MicroProfile specification. MicroProfile Context Propagation enables you to create completion stages that behave deterministically with respect to thread context and leverages the autonomic tuning of the Liberty global thread pool for asynchronous dependent stages.

Reviewer: ```suggestion
MicroProfile Context Propagation is a standalone MicroProfile specification. This new Open Liberty feature enables you to create completion stages that behave deterministically with respect to thread context and leverages the autonomic tuning of the Liberty global thread pool for asynchronous dependent stages.
```

Author: Done

----

Example usage within a Servlet:

[source, java]
----
private ManagedExecutor executor;

public void init(ServletConfig config) throws ServletException {
    executor = ManagedExecutor.builder()
                .propagated(ThreadContext.APPLICATION)
                .cleared(ThreadContext.ALL_REMAINING)
                .build();
}

public void destroy() {
    executor.shutdownNow();
}

public void doGet(HttpServletRequest req, HttpServletResponse resp)
    throws ServletException, IOException {
    ...
    executor.copy(unmanagedCompletionStage).thenAcceptAsync(value -> {
        // requires java:comp namespace of the application,
        DataSource ds = InitialContext.doLookup("java:comp/env/jdbc/ds");
        ...
    });
}
----

Find out more: 

* link:https://download.eclipse.org/microprofile/microprofile-context-propagation-1.2-RC1/microprofile-context-propagation-spec-1.2-RC1.html[The MicroProfile Context Propagation 1.2 Release Candidate 1 specification]
* link:https://download.eclipse.org/microprofile/microprofile-context-propagation-1.2-RC1/apidocs/[The MicroProfile Context Propagation 1.2 JavaDoc]

[#leakedConnections]
=== Automatic Cleanup of Leaked Connections

> Liberty connection management is enhanced with the ability to automatically detect and close unsharable connections that are left open by the application across the end of a request.

Reviewer: ```suggestion
Open Liberty connection management is enhanced with the ability to automatically detect and close unsharable connections that are left open by the application across the end of a request.
```

Author: Done

> Occasionally, application code might forget to close an unsharable connection that it obtains, which prevents the connection from being returned to the connection pool for use by other requests. Over time, these leaked connections can degrade performance and eventually exhaust the connection pool.  Liberty connection management now has the ability to detect and automatically close these leaked connections to prevent this from happening.

Reviewer: ```suggestion
Occasionally, application code might forget to close an unsharable connection that it obtains. This prevents the connection from being returned to the connection pool for use by other requests. Over time, these leaked connections can degrade performance and eventually exhaust the connection pool. Open Liberty connection management now has the ability to detect and automatically close these leaked connections to prevent this from happening.
```

Author: Done

> To take advantage of this new capability, configure one of the Liberty features that leverages the `connectionManager` element. For example, JDBC:

Reviewer: ```suggestion
To take advantage of this new capability, configure one of the Open Liberty features that leverages the `connectionManager` element. For example, JDBC:
```

Author: Done

:imagesdir: /
:url-prefix:
:url-about: /

'''

We have two beta packages for link:{url-about}[Open Liberty]:

* <<allbeta, All Beta Features>>: a larger package that contains all Open Liberty beta features (including Jakarta EE 9 beta features) and GA features and functions.
* <<jakarta, Jakarta EE 9 Beta Features>>: a lightweight package that contains only the Jakarta EE 9 features.

This means that you can now try out our in-development Open Liberty features by just adding the relevant coordinates to your build tools.

If you try either package, <<feedback, let us know what you think>>.

In addition to all Jakarta EE 9 and MicroProfile 4.0 beta features, Open Liberty 21.0.0.3-beta also contains five brand new features.

'''

[#allbeta]
== All Beta Features package

The All Beta Features package includes the following beta features:

* <<MPContext, MicroProfile Context Propagation 1.2>>
* <<requestTiming, Ability to control whether thread dumps are collected when a hung request is detected in the Request Timing feature>>
* <<leakedConnections, Automatic Cleanup of Leaked Connections>>
* <<expandLocation, expandLocation property for App Manager configuration>>

> These features also join MicroProfile 4.0 features which can be found link:https://openliberty.io/blog/2021/01/26/ee9-messaging-security-21002-beta.html#allbeta[here].

Reviewer 2: This is not quite straightforward as the previous beta blog removed MicroProfile Fault Tolerance 3.0. The 21.0.0.3-beta has the full MP 4.0 features. I suggest listing all of lists to make it complete. Besides, we would like to use this beta blog for GA blog.

Author: Done

----
<requestTiming includeContextInfo="true" slowRequestThreshold="120s" hungRequestThreshold="10s" sampleRate="1" enableThreadDumps="false"></requestTiming>`
----

The `enableThreadDumps` server configuration attribute can also be used in embedded Request Timing sub-elements: 
`<servletTiming/>` or `<jdbcTiming/>`, as follows:

[source, xml]
----
<requestTiming includeContextInfo="true" slowRequestThreshold="120s" hungRequestThreshold="10s" sampleRate="1">
    <servletTiming appName="MyApp" servletName="MyServletApp" slowRequestThreshold="100s" hungRequestThreshold="5s" enableThreadDumps="false"/>
</requestTiming>`
----

Note: An embedded `<servletTiming/>` or `<jdbcTiming/>` configuration element in the server.xml file overrides the configured slow and hung request threshold that are defined in `<requestTiming/>`. 

For more information on the Request Timing feature, please refer to the following documentations:
- link:{url-prefix}/docs/21.0.0.1/reference/feature/requestTiming-1.0.html[Open Liberty Documentation on requestTiming-1.0 feature]
- link:{url-prefix}/docs/21.0.0.1/reference/config/requestTiming.html[Open Liberty Documentation on requestTiming Configuration]

'''

> [MPContext]

Reviewer 2: ```suggestion
[#MPContext]
```

Author: Good catch! Thank you!

blog_description: In addition to all Jakarta EE 9 and MicroProfile 4.0 beta features, Open Liberty 21.0.0.3-beta also contains five brand new features.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= MicroProfile Context Propagation 1.2, RequestTiming 1.0 and JakartaMail 2.0 in Open Liberty 21.0.0.3-beta
Jakub Pomykala <https://github.com/jakub-pomykala>
:imagesdir: /
:url-prefix:
:url-about: /

'''

We have two beta packages for link:{url-about}[Open Liberty]:

* <<allbeta, All Beta Features>>: a larger package that contains all Open Liberty beta features (including Jakarta EE 9 beta features) and GA features and functions.
* <<jakarta, Jakarta EE 9 Beta Features>>: a lightweight package that contains only the Jakarta EE 9 features.

This means that you can now try out our in-development Open Liberty features by just adding the relevant coordinates to your build tools.

> 

Reviewer: Insert a "Try it now" link in here to link to the section at the bottom

Author: Done

  <feature>mpContextPropagation-1.2</feature>
  <!-- other features used by example code... -->
  <feature>servlet-4.0</feature>
  <feature>jdbc-4.2</feature>
  <feature>jndi-1.0</feature>
</featureManager>
----

Example usage within a Servlet:

[source, java]
----
private ManagedExecutor executor;

public void init(ServletConfig config) throws ServletException {
    executor = ManagedExecutor.builder()
                .propagated(ThreadContext.APPLICATION)
                .cleared(ThreadContext.ALL_REMAINING)
                .build();
}

public void destroy() {
    executor.shutdownNow();
}

public void doGet(HttpServletRequest req, HttpServletResponse resp)
    throws ServletException, IOException {
    ...
    executor.copy(unmanagedCompletionStage).thenAcceptAsync(value -> {
        // requires java:comp namespace of the application,
        DataSource ds = InitialContext.doLookup("java:comp/env/jdbc/ds");
        ...
    });
}
----

Find out more: 

* link:https://download.eclipse.org/microprofile/microprofile-context-propagation-1.2-RC1/microprofile-context-propagation-spec-1.2-RC1.html[The MicroProfile Context Propagation 1.2 Release Candidate 1 specification]

> * link:https://download.eclipse.org/microprofile/microprofile-context-propagation-1.2-RC1/apidocs/[The MicroProfile Context Propagation 1.2 JavaDoc]

Reviewer: ```suggestion
* link:https://download.eclipse.org/microprofile/microprofile-context-propagation-1.2-RC1/apidocs/[The MicroProfile Context Propagation 1.2 Release Candidate 1 JavaDoc]
```

Author: Done

To take advantage of this new capability, configure one of the Open Liberty features that leverages the `connectionManager` element. For example, JDBC:

[source, xml]
----
<featureManager>
  <feature>jdbc-4.2</feature>
  <feature>jndi-1.0</feature>
  <!-- more features -->
</featureManager>
----

Configure connection managers for your data sources to enable the new `autoCloseConnections` attribute,

[source, xml]
----
<dataSource id="DefaultDataSource">
  <connectionManager maxPoolSize="10" autoCloseConnections="true"/>
    <jdbcDriver libraryRef="PostgreSQL"/>
    <properties.postgresql databaseName="TESTDB" serverName="localhost" portNumber="5432"/>
</dataSource>

<library id="PostgreSQL">
    <file name="/usr/local/postgresql/postgresql-42.2.18.jar"/>
</library>
----

Find out more:

* link:{url-prefix}/docs/21.0.0.3/reference/config/connectionManager.html[connectionManager config documentation]

'''

[#jakarta]
== Jakarta EE 9 Beta Features package

> The main change visible to developers in Jakarta EE is the names of packages changing to accomodate the new `jakarta.*` namespace. In this Open Liberty beta, we have a number of new API Release Candidates to join the expanding library of supported Jakarta packages.

Reviewer: I thought we dropped this paragraph?

Author: Yes we did, removed again.

> This Open Liberty beta introduces the following Jakarta EE 9 feature which now possess it's all-new Jakarta EE 9 package name:

Reviewer: ```suggestion
This Open Liberty beta introduces the following Jakarta EE 9 feature which now possesses its all-new Jakarta EE 9 package name:
```

Author: Done

Configuring mail sessions works basically the same as with the Liberty Feature JavaMail-1.5 and JavaMail-1.6. They can be configured using the API, or through the server.xml 

Below is an example of a SMTP Mail session configured through the server.xml:

[source, xml]
----
<featureManager>
  <feature>mail-2.0</feature>
  <feature>jndi-1.0</feature>
</featureManager>

<mailSession>
      <mailSessionID>testSMTPMailSession</mailSessionID>
      <jndiName>TestingApp/SMTPMailSessionServlet/testSMTPMailSession</jndiName>
      <description>mailSession for testing SMTP protocol</description>
      <transportProtocol>smtp</transportProtocol>
      <host>localhost</host>
      <user>somuser@someemailserver.com</user>
      <password>usersPassword</password>
      <from>someuser@someemailserver.com</from>
      <property name="mail.smtp.host" value="localhost" \>
      <property name="mail.smtp.port" value="3025" \>
  </mailSession> 
----

Find out more:

* link:https://eclipse-ee4j.github.io/mail/[Jakarta mail]

> * link:https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_admin_javamail.html[Administering on liberty substitute JavaMail-1.5 with mail-2.0]

Reviewer: I think we agreed to remove this link ... but did you want to ask the developer first?

Author: Waiting for the dev to get back to me. Will check with him again today.

Author: Done

> Enable the Jakarta EE 9 beta features in your app's `server.xml`. You can enable the individual features you want or you can just add the Jakarta EE 9 convenience feature to enable all of the Jakarta EE 9 beta features at once:

Reviewer: ```suggestion
Enable the Jakarta EE 9 beta features in your server's `server.xml` configuration file. You can enable the individual features you want or you can just add the Jakarta EE 9 convenience feature to enable all of the Jakarta EE 9 beta features at once:
```

Author: Done
